### PR TITLE
issue #38 v2: implementing token-based error location

### PR DESCRIPTION
**Pull request by ntwrick**
Well, it turns out that any lexer.CurrentPosition() based approach is useless for error location because the lexer fast forward functions such as lexer.skipWhiteSpace(), lexer.readLine(), et al. change the current position before the interpreter reports the error. 

So, the new model does token-based position marking which is totally accurate (i.e. hack-free). In this model the lexer captures the current lexer.position in lexer.NextToken() (before it can move away) and saves it into the new Token.Position field. Since the parser already places the Token into each of the existing ast.Node.Token fields, we carry the token.Position along for the ride into the evaluator. And, as before, we pass the lexer pointer thru to the evaluator, but the REPL and testing modules now call a new BeginEval(node, *env, *lexer) function which, in turn, calls the recursive Eval(node, *env) function.

Thus, when we hit either a parser or evaluator error, their error location reporters pull the token.Location and pass it to the lexer.ErrorLine(pos) method which does the position look-up in the lexer.lineMap and returns the lineNum, column, and errorLine for formatting into the error location message.

Bottom line is that both the line numbers and columns are now accurate in both the parser and evaluator error messages without requiring any backtracking in the file.

I will withdraw the previous PR and, upon further review, we can go forward with this PR v2 instead.

### CODE REVIEW

File: evaluator/evaluator_test.go [hunk 1]
```diff
@@ -660,7 +651,7 @@ func TestBuiltinFunctions(t *testing.T) {
 		{`find([1,2], f(x) {x == "some"})`, nil},
 		{`arg("o")`, "argument 0 to arg(...) is not supported (got: o, allowed: NUMBER)"},
 		{`arg(3)`, ""},
-		{`pwd().split("").reverse().slice(0, 33).reverse().join("").replace("\\", "/", -1)`, "github.com/abs-lang/abs/evaluator"}, // Little trick to get travis to run this test, as the base path is not /go/src/
+		{`pwd().split("").reverse().slice(0, 33).reverse().join("").replace("\\", "/", -1)`, "/evaluator"}, // Little trick to get travis to run this test, as the base path is not /go/src/
```
Comment (odino) [inline evaluator/evaluator_test.go]: I'm a little bit worried about changing all tests to `Contains`. Maybe it's worth isolating, or commenting this test out.

File: object/object.go [hunk 2]
```diff
@@ -112,9 +120,9 @@ func (f *Function) Inspect() string {
 	out.WriteString("f")
 	out.WriteString("(")
 	out.WriteString(strings.Join(params, ", "))
-	out.WriteString(") {\n")
+	out.WriteString(") {")
```
Comment (odino) [inline object/object.go]: why removing the new lines?

File: parser/parser.go [hunk 3]
```diff
@@ -358,7 +364,8 @@ func (p *Parser) curPrecedence() int {
 
 // var
 func (p *Parser) parseIdentifier() ast.Expression {
-	return &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
+	id := &ast.Identifier{Token: p.curToken, Value: p.curToken.Literal}
+	return id
```
Comment (odino) [inline parser/parser.go]: why 2 lines here?

File: parser/parser.go [hunk 4]
```diff
@@ -477,7 +486,8 @@ func (p *Parser) parseMethodExpression(object ast.Expression) ast.Expression {
 
 // true
 func (p *Parser) parseBoolean() ast.Expression {
-	return &ast.Boolean{Token: p.curToken, Value: p.curTokenIs(token.TRUE)}
+	b := &ast.Boolean{Token: p.curToken, Value: p.curTokenIs(token.TRUE)}
+	return b
```
Comment (odino) [inline parser/parser.go]: same here

File: evaluator/evaluator_test.go (hunk 1 above)
Comment (ntwrick) [reply evaluator/evaluator_test.go]: There are at least a couple of cases where I think that the `error == expected` testing paradigm does not work well: 

1) environment dependent tests such as `pwd()` and `$(command)` can be retained as long as we match the  important parts of the results, and
2) tests that contain error location lines that are not essential to the test result.

By using `strings.Contains(error, expected)` we can focus the match on the critical part of the result or even the entire result if we wish. But if we always use `error == expected` we are limiting our flexibility.

File: object/object.go (hunk 2 above)
Comment (ntwrick) [reply object/object.go]: This is related to the `error == expected` testing problem.  `object.Function.Inspect()` was returning a result containing gratuitous newlines that complicate matching the expected result in the tests. I don't understand why the internal representation of a function should contain newlines (newlines are for people, not for parsers :-)). I also don't see that there is any other user of `Inspect()` other than the testing module and so `Function.Inspect()` seems to be the only one doing this kind of formatted result string.

File: parser/parser.go (hunk 3 above)
Comment (ntwrick) [reply parser/parser.go]: Fixed this. I may have been reacting to a complaint from my aggressive VScode go-lint extension. I love VScode, its a great IDE, but sometimes more of a nanny :-)). 

In particular it complains about unexported external functions that don't have canonical comments preceding the func def. See https://blog.golang.org/godoc-documenting-go-code. But I only did this for new funcs that I created, so there are still a few left in the code. Priorities ... Priorities...

**Comment by odino** [comment]
Thanks a ton! :fireworks: :fireworks: :fireworks:
